components/service/glean/src/main/java/mozilla/components/service/glean/StringMetricType.kt

```diff
@@ -0,0 +1,59 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean
+
+import mozilla.components.service.glean.storages.StringsStorageEngine
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This implements the developer facing API for recording string metrics.
+ *
+ * Instances of this class type are automatically generated by the parsers at build time,
+ * allowing developers to record values that were previously registered in the metrics.yaml file.
+ *
+ * The string API only exposes the [record] method, which takes care of validating the input
+ * data and making sure that limits are enforced.
+ */
+data class StringMetricType(
+    override val applicationProperty: Boolean = false,
```

Comment:
Mh, I think I found a Kotlin way to have defaults declared on the `CommonMetricData` interface. According to [the docs](https://kotlinlang.org/docs/reference/interfaces.html#properties-in-interfaces), we can do something like this:

```
interface CommonMetricData {
   ...
    val disabled: Boolean?
        get() = false
    val category: String?
    ....
}
```

Since we're always using the same defaults for each metric type, maybe we should just refactor this right now? What do you think? Do you want to address that in this PR or should we file a follow-up and take it there, before we use other types?

Reply:
Thanks for finding that.  I actually looked and didn't come across this in my searching.  Let's try to do this now -- it will make adding more metric types easier going forward.

Reply:
Trying this, I'm not sure this makes things any better.  Putting the `get() = false` in the interface indeed sets default behavior for the interface, but you still have to override it in the data class so that you can give kotlin a place to store the actual value for the case where it's different from the default.  And you still need to provide the default in the data class to generate a default value for the constructor's arguments.  So in short, I don't think making this change is useful for us unless there were specific metric types that wouldn't need to touch / store values for these fields (which I don't think will ever be the case).

This leads me back to my earlier thought: we could have `glean_parser` *always* write out the values for arguments, even if they are the default.  Then we wouldn't have to declare the defaults in each metric type implementation -- the defaults would only be stored and applied by the glean_parser when loading a `metrics.yaml` file...

What do you think?

Reply:
Oh, good point. I like the idea of the parsers always writing values for arguments. But let's tackle this right after this PR is merged and not block on it. Meanwhile, let's leave the defaults as you did initially.

Reply:
Sounds like a plan.